### PR TITLE
Kafka connect

### PR DESCRIPTION
**Pull request by meiao**
### Overview
Adds instrumentation for Kafka Connect.
There are 3 new instrumentation modules.
The first adds support for extracting metrics from Kafka Connect.
The other 2 add transaction tracing, one for versions 2.0.0 up to 3.2.3, and the other for versions 3.3.0 and later.

### CODE REVIEW

File: instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                this::harvest, 0L, REPORTING_INTERVAL_IN_SECONDS, TimeUnit.SECONDS);
+    }
+
+    // default visibility for testing
+    void harvest() {
+        try {
+            Map<String, Object> eventData = METRICS_AS_EVENTS ? new HashMap<>() : Collections.emptyMap();
+            for (final Map.Entry<String, KafkaMetric> metric : metrics.entrySet()) {
+                Object metricValue = metric.getValue().metricValue();
+                if (metricValue instanceof Double) {
```
Comment (obenkenobi) [inline instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java:78]: Why not use `metricValue instanceof Number`? There could be missing metrics that cast to an integer.

Comment (obenkenobi) [reply instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java:78]: In addition you can then get your float value like this: `final float value = ((Number) metricValue).floatValue();`

File: instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    private static final AtomicBoolean INITIALIZED = new AtomicBoolean(false);
+
+    // key is the metric/event name
+    private final Map<String, KafkaMetric> metrics = new ConcurrentHashMap<>();
+
+    private ScheduledFuture<?> scheduledFuture;
+
+    // the metrics are shared between Workers. So only one reporter needs to be attached.
+    public static void initialize(ConnectMetrics.MetricGroup metricGroup) {
```
Comment (obenkenobi) [inline instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java:51]: Nice. I'd like to refactor some of the other Kafka metrics reporters in streams/clients to use this approach.

**Review by obenkenobi — APPROVED** [review]

**Review by jtduffy — APPROVED** [review]
+1

File: instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java (hunk 1 above)
Comment (meiao) [reply instrumentation/kafka-connect-metrics-1.0.0/src/main/java/com/nr/instrumentation/kafka/connect/KafkaConnectMetricsReporter.java:78]: This was copied from the original Kafka metrics instrumentation.
But that is a good change. I checked and apparently they do have integer metrics.

**Pull request merged**